packages/astro/src/content/internal.ts

```diff
@@ -69,18 +69,28 @@ export function createGetCollection({
 	};
 }
 
-export function createGetEntry({
-	collectionToEntryMap,
+export function createGetEntryBySlug({
+	getCollection,
 	collectionToRenderEntryMap,
 }: {
-	collectionToEntryMap: CollectionToEntryMap;
+	getCollection: ReturnType<typeof createGetCollection>;
 	collectionToRenderEntryMap: CollectionToEntryMap;
 }) {
-	return async function getEntry(collection: string, entryId: string) {
-		const lazyImport = collectionToEntryMap[collection]?.[entryId];
-		if (!lazyImport) throw new Error(`Failed to import ${JSON.stringify(entryId)}.`);
+	return async function getEntryBySlug(collection: string, slug: string) {
+		const entries = await getCollection(collection);
```

Comment:
nit: might be good to add a comment/todo to improve to O(1) lookup in the future, just for any sleuths (or future contributors) who find this code and wonder why it was implemented this way.